amethyst_rendy/src/bundle.rs

```diff
@@ -0,0 +1,943 @@
+pub trait RenderPlugin<B: Backend>: std::fmt::Debug {
+    /// Hook for adding systems and bundles to the dispatcher.
+    fn build<'a, 'b>(&mut self, _builder: &mut DispatcherBuilder<'a, 'b>) -> Result<(), Error> {
+        Ok(())
+    }
+
+    /// Hook for providing triggers to rebuild the render graph.
+    fn rebuild(&mut self, _res: &Resources) -> bool {
+        false
+    }
+
+    /// Hook for extending the rendering plan.
+    fn plan(
+        &mut self,
+        plan: &mut RenderPlan<B>,
+        factory: &mut Factory<B>,
+        res: &Resources,
+    ) -> Result<(), Error>;
+}
+
+/// Builder of a rendering plan for specfied target.
+#[derive(Debug)]
+pub struct RenderPlan<B: Backend> {
+    targets: HashMap<Target, TargetPlan<B>>,
+    roots: Vec<Target>,
+}
+
+impl<B: Backend> RenderPlan<B> {
+    fn new() -> Self {
+        Self {
+            targets: Default::default(),
+            roots: vec![],
+        }
+    }
+
+    /// Mark rander target as root. Root render targets are always
+    /// evaluated, even if nothing depends on them.
+    pub fn add_root(&mut self, target: Target) {
+        if !self.roots.contains(&target) {
+            self.roots.push(target);
+        }
+    }
+
+    /// Define a render target with predefined set of outputs.
+    pub fn define_pass(
+        &mut self,
+        target: Target,
+        outputs: TargetPlanOutputs<B>,
+    ) -> Result<(), Error> {
+        let target_plan = self
+            .targets
+            .entry(target)
+            .or_insert_with(|| TargetPlan::new(target));
+
+        target_plan.set_outputs(outputs)?;
+
+        Ok(())
+    }
+
+    /// Extend the rendering plan of a render target. Target can be defined in other plugins.
+    /// The closure is evaluated only if the target contributes to the rendering result, e.g.
+    /// is rendered to a window or is a dependency of other evaluated target.
+    pub fn extend_target(
+        &mut self,
+        target: Target,
+        closure: impl FnOnce(&mut TargetPlanContext<'_, B>) -> Result<(), Error> + 'static,
+    ) {
+        let target_plan = self
+            .targets
+            .entry(target)
+            .or_insert_with(|| TargetPlan::new(target));
+        target_plan.add_extension(Box::new(closure));
+    }
+
+    fn build(self, factory: &Factory<B>) -> Result<GraphBuilder<B, Resources>, Error> {
+        let mut ctx = PlanContext {
+            target_metadata: self
+                .targets
+                .iter()
+                .filter_map(|(k, t)| unsafe { t.metadata(factory.physical()) }.map(|m| (*k, m)))
+                .collect(),
+            targets: self.targets,
+            passes: Default::default(),
+            outputs: Default::default(),
+            graph_builder: GraphBuilder::new(),
+        };
+
+        for target in self.roots {
+            // prevent evaluation of roots that were accessed recursively or undefined
+            if !ctx.passes.contains_key(&target) {
```

Comment:
This check should be moved inside the `evaluate_target` function, imo. It's not immediately obvious just from this snippet that evaluating the target can add it to the list of passes.